drivers/flash/flash_qspi_renesas_ra.c

```diff
@@ -0,0 +1,501 @@
+	.pages_size = QSPI_ERASE_BLK_SZ,
+};
+
+void qspi_flash_ra_page_layout(const struct device *dev, const struct flash_pages_layout **layout,
+			       size_t *layout_size)
+{
+	ARG_UNUSED(dev);
+	*layout = &qspi_flash_ra_layout;
+	*layout_size = 1;
+}
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+#if defined(CONFIG_FLASH_JESD216_API)
+static int qspi_flash_ra_read_jedec_id(const struct device *dev, uint8_t *id)
+{
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+	int err = 0;
+	uint8_t cmd;
+
+	if (id == NULL) {
+		return -EINVAL;
+	}
+	acquire_device(dev);
+
+	if (SPI_FLASH_PROTOCOL_QPI == qspi_data->qspi_cfg.spi_protocol) {
+		cmd = QSPI_CMD_READ_ID;
+	} else {
+		cmd = Ex_SPI_CMD_READ_ID;
+	}
+	err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, true);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: direct write failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = R_QSPI_DirectRead(&qspi_data->qspi_ctrl, id, THREE_BYTE);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: direct read failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = get_flash_status(dev);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: failed to get status for QSPI operation");
+		err = -EIO;
+	}
+out:
+	release_device(dev);
+
+	return err;
+}
+
+static int qspi_flash_ra_sfdp_read(const struct device *dev, off_t addr, void *data, size_t size)
+{
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+	int err = 0;
+	uint8_t buffer[size];
+
+	acquire_device(dev);
+	memset(&buffer[0], 0, sizeof(buffer));
+	buffer[0] = QSPI_CMD_READ_SFDP;
+	buffer[1] = addr;
+	buffer[2] = addr >> 8;
+	buffer[3] = addr >> 16;
+
+	err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &buffer[0], FOUR_BYTE, true);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: direct write failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = R_QSPI_DirectRead(&qspi_data->qspi_ctrl, &buffer[0], size);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: direct read failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = get_flash_status(dev);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: failed to get status for QSPI operation");
+		err = -EIO;
+		goto out;
+	}
+
+	if (SPI_FLASH_PROTOCOL_QPI == qspi_data->qspi_cfg.spi_protocol) {
+		/* 3 dummy byte */
+		memcpy(data, &buffer[4], size);
+	} else {
+		/* 1 dummy byte */
+		memcpy(data, &buffer[1], size);
+	}
+out:
+	release_device(dev);
+	return err;
+}
+#endif
+
+static bool qspi_flash_ra_valid(off_t area_size, off_t offset, size_t len)
+{
+	if ((offset < 0) || (offset >= area_size) || ((area_size - offset) < len)) {
+		return false;
+	}
+
+	return true;
+}
+
+static int qspi_flash_ra_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+	int err = 0;
+	struct flash_pages_info page_info_start, page_info_end;
+	uint32_t erase_size;
+	int rc;
+
+	if (!len) {
+		return 0;
+	}
+
+	if (!qspi_flash_ra_valid(QSPI_NOR_FLASH_SIZE, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len % QSPI_ERASE_BLK_SZ != 0) {
+		return -EINVAL;
+	}
```

Comment:
You probably want to have some logging in at least one of the ifs, otherwise you have the same error returned to user with no way to distinguish problems except for debugging.

Reply:
We fixed it. We added log to distinguish for errors.